### PR TITLE
Remove Req/Res Meta

### LINKED ISSUE #617: Discuss removing req/res meta
What is the API that users expect? I'll bet it's something like this:

```go
resp, err := client.Call(ctx, req)
```

and for handlers:

```go
func (handler) Call(ctx context.Context, e *service.Request) (*service.Response, error) {
	return &service.Response{}, nil
}
```

But what about metadata?

### Outbound Metadata

For outbound metadata, like headers, I think that riding the context is a fine option:

```go
resp, err := client.Call(yarpc.WithHeaders(ctx, {"blah": "blah"}), req)

func (handler) Call(ctx context.Context, e *service.Request) (*service.Response, error) {
        headers := yarpc.GetHeaders(ctx)
	return &service.Response{}, nil
}
```

### Inbound Metadata

For inbound metadata, we likely need to return *something* from the handler:

```go
func (handler) Call(ctx context.Context, e *service.Request) (*service.Response, yarpc.ResMeta, error) {
	return &service.Response{}, nil, nil
}
```

And to receive inbound metadata at outbound callsites, we can use the options pattern:

```go
resMeta := make(yarpc.ResMeta)
resp, err := client.Call(ctx, req, yarpc.WithResMeta(resMeta))
```

Alot of this is pseudo-code, but I just wanted to get some ideas out there.

### PR DESCRIPTION
**Pull request by abhinav**
This PR removes Req/ResMeta types completely.

Call site options are now provided via `CallOption` varargs. Incoming request
information is now accessed using the `CallFromContext` API.

All commits are individually reviewable.

- [x] Client-side APIs #648 
- [x] Server-side APIs #649
- [x] Raw client #647
- [x] Raw server #650
- [x] JSON client #652
- [x] JSON server #651
- [x] Thrift client #654
- [x] Thrift server #657
- [x] Documentation updates

Resolves #617

### CODE REVIEW

**Comment by abhinav** [comment]
@breerly, @kriskowal: @prashantv thinks we should move the accessors into a
single interface:

```
call := yarpc.CurrentCall(ctx)
fmt.Println(call.Caller())
fmt.Println(call.Header("foo"))
call.WriteResponseHeader("foo", "bar")
```

This has the advantage of discoverability in godocs. Right now, all
`CallOption`s will be listed together because they're the same type but
`context.Context` accessors will not because they're just normal functions.
Putting them in an interface together will make them all visible.

Thoughts?

**Comment by breerly** [comment]
Either that or move all the top-level funcs to their own package. If we go the single-shape route, let's be sure we can append new props to it, e.g. return struct instead of interface.

Don't feel super strongly either way.

**Comment by kriskowal** [comment]
Fine by me. Please consider `yarpc.CallFromContext(ctx)` as in `opentracing.SpanFromContext(ctx)`.

**Comment by breerly** [comment]
+1 to `yarpc.CallFromContext`

**Comment by abhinav** [comment]
CallFromContext: https://github.com/yarpc/yarpc-go/pull/658

File: call.go [hunk 1]
```diff
@@ -130,3 +133,170 @@ func WithRoutingKey(rk string) CallOption {
 func WithRoutingDelegate(rd string) CallOption {
 	return CallOption{func(o *OutboundCall) { o.routingDelegate = &rd }}
 }
+
+// InboundCall is an incoming call. It holds information about the inbound
```
Comment (breerly) [inline call.go]: Just "InboundCall holds information..." is likely enough

File: call.go [hunk 2]
```diff
@@ -130,3 +133,170 @@ func WithRoutingKey(rk string) CallOption {
...
+// call and its response.
```
Comment (breerly) [inline call.go]: "it's" right?

Comment (breerly) [inline call.go]: Nm, just for "it is" or "it has".

File: call.go [hunk 3]
```diff
@@ -130,3 +133,170 @@ func WithRoutingKey(rk string) CallOption {
...
+//
+// Encoding authors may use InboundCall to provide information about the
+// incoming request on the Context and receive response headers through
```
Comment (breerly) [inline call.go]: "and send response headers"*

File: call.go [hunk 4]
```diff
@@ -130,3 +133,170 @@ func WithRoutingKey(rk string) CallOption {
...
+}
+
+type inboundCallKey struct{} // context key for *InboundCall
+
+// NewInboundCall builds a new InboundCall with the given context.
+//
+// A request context is returned and must be used in place of the original.
+func NewInboundCall(ctx context.Context) (context.Context, *InboundCall) {
+	call := &InboundCall{}
+	return context.WithValue(ctx, inboundCallKey{}, call), call
```
Comment (breerly) [inline call.go]: Using an unexported key has the effect of no other package being able to access this key?

File: call.go [hunk 5]
```diff
@@ -130,3 +133,170 @@ func WithRoutingKey(rk string) CallOption {
...
+	ic.req = req
+	return nil
+}
+
+// WriteToResponse writes response information from the InboundCall onto the
+// given ResponseWriter.
+//
+// If used, this must be called before writing the response body to the
+// ResponseWriter.
+func (ic *InboundCall) WriteToResponse(resw transport.ResponseWriter) error {
```
Comment (breerly) [inline call.go]: Is this automatic or do users have to call this before returning? If it's automatic, then maybe this should start out unexported?

File: call.go (hunk 4 above)
Comment (abhinav) [inline call.go]: Yep

File: call.go (hunk 5 above)
Comment (abhinav) [inline call.go]: It's not automatic. The encoding must call this explicitly to write response headers for non-oneway responses.

Comment (breerly) [inline call.go]: sgtm

**Comment by breerly** [comment]
No squash?

**Comment by abhinav** [comment]
Not for this. Each commit in this PR is self-contained with no "noise" commits. Easier to look at N self-contained commits rather than a thousand line squashed commit.

**Pull request merged**